apisix/plugins/ai-request-rewrite.lua

```diff
@@ -0,0 +1,222 @@
+        auth = auth_schema,
+        options = model_options_schema,
+        timeout = {
+            type = "integer",
+            minimum = 1,
+            maximum = 60000,
+            default = 30000,
+            description = "Total timeout in milliseconds for requests to LLM service, " ..
+            "including connect, send, and read timeouts."
+        },
+        keepalive = {
+            type = "boolean",
+            default = true
+        },
+        keepalive_pool = {
+            type = "integer",
+            minimum = 1,
+            default = 30
+        },
+        ssl_verify = {
+            type = "boolean",
+            default = true
+        },
+        override = {
+            type = "object",
+            properties = {
+                endpoint = {
+                    type = "string",
+                    description = "To be specified to override " ..
+                    "the endpoint of the AI service provider."
+                }
+            }
+        }
+    },
+    required = {"prompt", "provider", "auth"}
+}
+
+local _M = {
+    version = 0.1,
+    priority = 1073,
+    name = plugin_name,
+    schema = schema
+}
+
+local function request_to_llm(conf, request_table, ctx)
+    local ai_driver = require("apisix.plugins.ai-drivers." .. conf.provider)
+
+    local extra_opts = {
+        endpoint = core.table.try_read_attr(conf, "override", "endpoint"),
+        query_params = conf.auth.query or {},
+        headers = (conf.auth.header or {}),
+        model_options = conf.options
+    }
+
+    local res, err, httpc = ai_driver:request(conf, request_table, extra_opts)
+    if err then
+        return nil, nil, err
+    end
+
+    local resp_body, read_err = res:read_body()
+    httpc:close()
+    if read_err then
+        return nil, nil, nil, read_err
+    end
+
+    return res, resp_body
+end
+
+
+local function parse_llm_response(res_body)
+    local response_table, err = core.json.decode(res_body)
+
+    if err then
+        return nil, "failed to decode llm response " .. ", err: " .. err
+    end
+
+    if not response_table.choices then
+        return nil, "'choices' not in llm response"
+    end
+
+    local message = response_table.choices[1].message
+    if not message then
+        return nil, "'message' not in llm response choices"
+    end
+
+    return message.content
+end
+
+
+function _M.check_schema(conf)
+    -- openai-compatible should be used with override.endpoint
+    if conf.provider == "openai-compatible" then
+        local override = conf.override
+
+        if not override or not override.endpoint then
+            return false, "override.endpoint is required for openai-compatible provider"
+        end
+    end
+
+    return core.schema.check(schema, conf)
+end
+
+
+function _M.access(conf, ctx)
+    local client_request_body, err = core.request.get_body()
+    if err then
+        core.log.error("failed to get request body: ", err)
+        return HTTP_BAD_REQUEST, err
```

Comment:
It doesn't seem appropriate to log an error level for a client-side error. I think it can be changed to info level.